70_FLIPFluids/gpuRadixSort.h

```diff
@@ -0,0 +1,470 @@
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+            tempDataBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = tempDataBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, tempDataBuffer.get());
+			updated = true;
+        }
+		uint32_t bufSize = sizeof(uint32_t) * numIterations * NUM_SORT_BINS;
+		if (!globalHistogramsBuffer || globalHistogramsBuffer->getSize() != bufSize)
+		{
+			video::IGPUBuffer::SCreationParams params = {};
+		    params.size = bufSize;
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT | IGPUBuffer::EUF_TRANSFER_DST_BIT;
+            globalHistogramsBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = globalHistogramsBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, globalHistogramsBuffer.get());
+			updated = true;
+		}
+		bufSize = sizeof(uint32_t) * (numWorkgroups * NUM_SORT_BINS + 1);
+		if (!partitionHistogramBuffer || partitionHistogramBuffer->getSize() != bufSize)
+		{
+			video::IGPUBuffer::SCreationParams params = {};
+		    params.size = bufSize;
+		    params.usage = IGPUBuffer::EUF_STORAGE_BUFFER_BIT;
+            partitionHistogramBuffer = m_device->createBuffer(std::move(params));
+
+		    video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = partitionHistogramBuffer->getMemoryReqs();
+		    reqs.memoryTypeBits &= m_device->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+
+		    auto bufMem = m_device->allocate(reqs, partitionHistogramBuffer.get());
+			updated = true;
+		}
+
+		return updated;
+    }
+
+    struct SSortParams
+    {
+        uint32_t numElements;
+        uint32_t bitShift;
+		uint32_t numWorkgroups;
+		uint32_t numThreadsPerGroup;
+    };
+
+	const uint32_t numThreadsPerGroup = 32;		// subgroup size
+	const uint32_t numIterations = 4;
+
+    smart_refctd_ptr<ILogicalDevice> m_device;
+
+    smart_refctd_ptr<IGPUComputePipeline> m_radixSortPipeline;
+    smart_refctd_ptr<IGPUComputePipeline> m_buildHistogramPipeline;
+	smart_refctd_ptr<IGPUComputePipeline> m_scanSpinePipeline;
+
+    smart_refctd_ptr<IDescriptorPool> sortDsPool;
+    std::array<smart_refctd_ptr<IGPUDescriptorSet>, 2> m_radixSortDs;
+	smart_refctd_ptr<IDescriptorPool> histogramDsPool;
+	std::array<smart_refctd_ptr<IGPUDescriptorSet>, 2> m_histogramDs;
+	smart_refctd_ptr<IDescriptorPool> m_scanSpineDsPool;
+	smart_refctd_ptr<IGPUDescriptorSet> m_scanSpineDs;
+
+    // buffers
+    smart_refctd_ptr<video::IGPUBuffer> paramsBuffer;
+    smart_refctd_ptr<video::IGPUBuffer> tempDataBuffer;
+	smart_refctd_ptr<video::IGPUBuffer> globalHistogramsBuffer;
+	smart_refctd_ptr<video::IGPUBuffer> partitionHistogramBuffer;
```

Comment:
its a very bad idea for utilities to own their scratch memory (instead of telling you how much they need and user providing), this is because if you start using them on multiple queues or even in the same frame without sync you get a race condition that you don't know about (whereas if you take such resources from the outside its always apparent you're responsible for external sync).

Same argument for descriptor sets.


One could ofc say that "well then, you make multiple instances of the utility for simultaneous users" , but then why compile the same shaders and create pipelines multiple times if they haven't changed?